Introducing DataChunk for DataPipes batching

VitalyFedyunin opened the pull request:
This is part of TorchArrow DF support preparation, separating it to multiple PRs to simplify review process.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62769 Fix clang_tidy script
* **#62768 Introducing DataChunk for DataPipes batching**

Differential Revision: [D30149090](https://our.internmc.facebook.com/intern/diff/D30149090)

VitalyFedyunin committed:
Introducing DataChunk for DataPipes batching

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Introducing DataChunk for DataPipes batching"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Introducing DataChunk for DataPipes batching"




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -68,13 +68,19 @@ def _apply(self, data, nesting_level):
         elif nesting_level > 0:
-            if not isinstance(data, list):
+            if isinstance(data, DataChunk):
+                result = data.__class__([self._apply(i, nesting_level - 1) for i in data.raw_iterator()])
```
nit: It's better to use `type(data)` than `data.__class__`.
Here is an example: https://stackoverflow.com/a/10633356

VitalyFedyunin replied:
👍🏻

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -25,6 +25,32 @@
 T = TypeVar('T')
 
 
+class DataChunk(object):
```
How about we make `DataChunk` inheriting `list`? Then, we only need to handle `isinstance(data, list)` and return `type(data)(result)`

VitalyFedyunin replied:
Not always, sometimes we need `raw_iterator` for DataFrames

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -179,7 +181,8 @@ def __iter__(self) -> Iterator[List[T_co]]:
             bucket: List[T_co]
             batch: List[T_co] = []
-            for bucket in self.bucket_ds:
+            for bucket_or_chunk in self.bucket_ds:
+                bucket = list(bucket_or_chunk)
                 # In-place sort within bucket
                 bucket.sort(key=self.sort_key)
```
If `DataChunk` inherits from `list` then no need to convert to list and run in-place sort. And, then converting list back to `DataChunk` is not needed.

VitalyFedyunin replied:
This is weird reason, but I would rather keep explicit conversion here, because it looks ugly and will convert it sooner in this case (while adding support for DF)

VitalyFedyunin committed:
Update on "Introducing DataChunk for DataPipes batching"

This is part of TorchArrow DF support preparation, separating it to multiple PRs to simplify review process.



[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Introducing DataChunk for DataPipes batching"

This is part of TorchArrow DF support preparation, separating it to multiple PRs to simplify review process.



[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30149090).
